rlkit/torch/sac/proto.py

```diff
@@ -54,90 +70,116 @@ def __init__(self,
+            self.z_means = torch.stack([p[0] for p in z_params])
+            self.z_vars = torch.stack([p[1] for p in z_params])
+        # sum rather than product of gaussians structure
         else:
-            new_z = torch.mean(new_z, dim=1)
-        self.z = new_z
+            self.z_means = torch.mean(params, dim=1)
 
-    def update_z(self, in_):
+    def sample_z(self):
+        if self.use_ib:
+            posteriors = [torch.distributions.Normal(m, torch.sqrt(s)) for m, s in zip(torch.unbind(self.z_means), torch.unbind(self.z_vars))]
+            z = [d.rsample() for d in posteriors]
+            self.z = torch.stack(z)
+        else:
+            self.z = self.z_means
+
+    def update_posterior(self, in_):
         '''
-        update current task embedding
-         - by running mean for prototypical encoder
+        update current q(z|c) with new data
+         - by adding to sum of natural parameters for prototypical encoder
          - by updating hidden state for recurrent encoder
         '''
-        z = self.z
         num_z = self.num_z
 
         # TODO this only works for single task (t == 1)
-        new_z = self.task_enc(in_)
-        if new_z.size(0) != 1:
+        up_params = self.task_enc(in_) # task x batch x feat
+        if up_params.size(0) != 1:
             raise Exception('incremental update for more than 1 task not supported')
         if self.recurrent:
-            z = new_z
+            up_mu = up_params[..., :self.latent_dim]
+            up_ss = F.softplus(up_params[..., self.latent_dim:])
+            self.z_means = up_mu[None]
+            self.z_vars = up_ss[None]
         else:
-            new_z = new_z[0] # batch x feat
-            num_updates = new_z.size(0)
-            for i in range(num_updates):
-                num_z += 1
-                z += (new_z[i][None] - z) / num_z
-        if self.use_ib:
-            z = self.information_bottleneck(z)
+            num_updates = up_params.size(1)
+            # only have one task here
+            up_params = up_params[0] # batch x feat
+            curr_mu = self.z_means[0]
+            curr_ss = self.z_vars[0]
+            if self.use_ib:
+                natural_z = torch.cat(_canonical_to_natural(curr_mu, curr_ss)) #feat
+                up_mu = up_params[..., :self.latent_dim]
+                up_ss = F.softplus(up_params[..., self.latent_dim:])
+                new_natural = torch.cat(_canonical_to_natural(up_mu, up_ss), dim=1) # batch x feat
+                new_natural = torch.sum(new_natural, dim=0)
+                natural_z += new_natural
+                m, s  = _natural_to_canonical(natural_z[:self.latent_dim], natural_z[self.latent_dim:]) # feat
+                self.z_means = m[None]
+                self.z_vars = s[None]
+            else:
+                for i in range(num_updates):
+                    num_z += 1
+                    curr_mu += (up_params[i] - curr_mu) / num_z
```

Comment:
this computation seems bugged. if it's actually the mean, then it should be 
(x + (n-1) * mu) / n?

Reply:
it's equivalent. written this way it's just factorized into the sum of the old value and the new part